quanto/library/ext/cpp/udqmm.cpp

```diff
@@ -0,0 +1,25 @@
+#include "udqmm.h"
+#include "unpack.h"
+
+#include <iostream>
+#include <torch/extension.h>
+
+using namespace std;
+
+torch::Tensor udqmm(torch::Tensor &input, torch::Tensor &weights, torch::Tensor& scale, torch::Tensor& zeropoint, int axis, int bits, torch::IntArrayRef orig_shape) {
+    torch::Tensor unpacked_weights = unpack(weights, bits);
+    torch::Tensor dq_output = (unpacked_weights.to(torch::kInt8) - zeropoint.to(torch::kInt8)) * scale;
```

Comment:
Ok to cast the unpacked_weigths since they are uint8 (we could also change the unpack method to do the cast there, making it part of the contract). We should only assert and err on zeropoint though, using TORCH_CHECK.

Reply:
It will probably be cleaner do the cast in the unpack. So that we have the same dtype for `unpacked_weights` and `zeropoint` ! LMK what you think. If yes, should we do that in another PR ?

Reply:
I thought about that a little bit and I am inclined to keep returning an unsigned int.
The unpacked weights really are unsigned integer in the range [0, 2**(bits -1)], and the
only reason for the cast is the subtraction of a signed int.
Now regarding the cast, I think we could simply drop it, as I realized pytorch is doing an implicit type promotion to `int16` because of the subtraction:
```python
>>> a = torch.tensor(255, dtype=torch.uint8).to('mps')
>>> b = torch.tensor(1, dtype=torch.int8).to('mps')
>>> a - b
tensor(254, device='mps:0', dtype=torch.int16)
```

Reply:
The reason why pytorch is doing that is probably because the first term might become negative, and change the result:
```python
>>> a.to(torch.int8) - b
tensor(-2, device='mps:0', dtype=torch.int8)
```
In our case we know the first term will stay positive during the cast, so it is safe, and avoids the materialization of a large `int16` intermediate tensor.